### PR TITLE
Initial changes on the Asset Renew and Cancel

### PR DESCRIPTION
**Pull request by pramodht**

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1,16 @@
+**/lwc/**/*.css
```
Comment (eclapham) [inline .eslintignore]: Why all the ignores?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,26 @@
...
+## Read All About It
+
+- [Salesforce Extensions Documentation](https://developer.salesforce.com/tools/vscode/)
+- [Salesforce CLI Setup Guide](https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_intro.htm)
+- [Salesforce DX Developer Guide](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_intro.htm)
+- [Salesforce CLI Command Reference](https://developer.salesforce.com/docs/atlas.en-us.sfdx_cli_reference.meta/sfdx_cli_reference/cli_reference.htm)
+
+# AssetRenew
+
+ To deploy this project first you have to enable the Subcription Management Licence and Assign all the required perms sets to access the APIs
```
Comment (eclapham) [inline README.md:26]: Is this required to even deploy this code? Or just to use it? Also, `license`.

File: config/project-scratch-def.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
+{
+  "orgName": "Subscription Management",
+  "edition": "Developer",
+  "features": ["EnableSetPasswordInApi"],
```
Comment (eclapham) [inline config/project-scratch-def.json]: Is there no way to create a SM org using features? Shouldn't customers be able to create scratch orgs using SM?

File: force-app/main/default/classes/AssetManagementController.cls [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/**
+ *
```
Comment (eclapham) [inline force-app/main/default/classes/AssetManagementController.cls]: What's up with empty comments? Shouldn't eslint catch this sort of thing?

File: force-app/main/default/classes/AssetManagementController.cls [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
+ * 
+ */
+public with sharing class AssetManagementController {
+
+    /**
+     * This is used for retreiving the Assets from Account which has HasLifecycleManagement is true
```
Comment (eclapham) [inline force-app/main/default/classes/AssetManagementController.cls:4]: Do we have a standard for comment formats or anything? I only ask because I don't know how we should portray "reference" implementations.

File: force-app/main/default/classes/AssetManagementControllerTest.cls [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        Asset asset = new Asset(Name = 'Testing', AccountId = acc.Id, Product2Id = product.Id);
+        insert asset;
+    }
+    
+    @isTest static void testRenewAsset() {
+        makeData();
+        
+        List<Asset> assetList = [Select Id from Asset limit 1];
+        List<String> responseList;
+        if(!assetList.isEmpty()){
```
Comment (eclapham) [inline force-app/main/default/classes/AssetManagementControllerTest.cls:41]: So if you fail to insert data your test just... passes silently?

File: force-app/main/default/classes/AssetManagementControllerTest.cls-meta.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>54.0</apiVersion>
```
Comment (eclapham) [inline force-app/main/default/classes/AssetManagementControllerTest.cls-meta.xml]: Shouldn't this match the controller itself, v55?

File: force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      // Wait for any asynchronous DOM updates
+      await flushPromises();
+
+      const tableEl = element.shadowRoot.querySelector('lightning-datatable');
+
+      //Validate the datatable is populated with correct number of records
+      expect(tableEl.data.length).toBe(getMockAssets.length);
+  });
+
+  it('Renew button is disabled', async () => {
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js]: Again, not sure what we consider "reference" quality, but your test names should be "readable". "it renew button is disabled" doesn't really tell me under what condition you expect the button to be disabled, etc.

File: force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+      });
+      document.body.appendChild(element);
+
+      // Emit data from @wire
+      getAssetsByAccount.emit(getMockAssets);
+      
+      // Wait for any asynchronous DOM updates
+      await flushPromises();
+
+      //Update multiple records with the INPUT_PARAMETERS and simulate the Save event
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js]: This comment seem inaccurate? You're just getting the button element from the DOM.

File: force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    const cancelButton = element.shadowRoot.querySelector('lightning-button[data-id=Cancel]');
+    dateEl.value = new Date().toJSON()
+    dateEl.dispatchEvent(new CustomEvent('change'));
+    // Wait for any asynchronous DOM updates
+    await flushPromises();
+    expect(cancelButton.disabled).toBe(false);
+    // Validate updateContacts call
+  });
+
+  it('When renew button is clicked', async () => {
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/__tests__/assetManagementRelated.test.js]: when clicked... what?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.html [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+<template>
+    <lightning-card title= "Asset Management Subscription">
+        <lightning-datatable key-field="Id" onrowselection = {handleRowSelection}
+                            data = {assetList} 
+                            columns = {columns}>
+        </lightning-datatable>
+        <c-pear-event-manager></c-pear-event-manager>
+        <div slot="actions">
+            <lightning-button-group>
+                <lightning-button variant="neutral" label="Renew Assets" name="Renew" class="slds-var-m-left_x-small" disabled = {activateButtons} onclick = {renewCancelAction}></lightning-button>
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.html]: Can you rearrange the properties so they're in the same order on both buttons, please? This will make them easier to compare.

Why does the renew button call `renewCancelAction`, is it renew or cancel, it can't be both. And the cancel button is a popup not an action?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
+import { LightningElement, wire, track, api} from 'lwc';
+import getAssets from '@salesforce/apex/AssetManagementController.getAssetsByAccount'
+import renewCancel from '@salesforce/apex/AssetManagementController.renewCancelAsset'
+const columns = [
+    { label: 'Asset Name', fieldName: 'Name', type: 'text' },
+    { label: 'Lifecycle Start Date', fieldName: 'LifecycleStartDate',type: 'date', typeAttributes: {
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js:8]: What happens when users don't have access to one or more of these fields, how does this component handle that?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        timeZone:"UTC"
+    }},
+    { label: 'Renewal Term Unit', fieldName: 'RenewalTermUnit', type: 'text'},
+    { label: 'Renewal Term', fieldName: 'RenewalTerm', type: 'Number'},
+];
+
+
+
+export default class AssetManagement extends LightningElement {
+    activateButtons = true;
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: I think a lot of your properties need better names. `activateButtons` sounds like a method name, an action you take. Not a boolean flag. Typically booleans are prepended with `is` or something similar, for example.

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        }
+    }
+
+    handleRowSelection = event => {
+        this.selectedRows =event.detail.selectedRows;
+        console.log(this.selectedRows);
+        this.activateButtons =  this.selectedRows.length > 0 ?  false : true;
+    }
+
+    renewCancelPopup = () => {
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: If this function toggles the popup, then name it such. `toggleRenewCancelPopup` or something similar.

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        this.showCancelPopup=  this.showCancelPopup === false ? true : false
+    }
+
+    renewCancelAction = event => {
+        let actionType = event.currentTarget.name;
+        //new Date(this.cancelledDate.toString().split('GMT')[0]+' UTC').toISOString()
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: If you don't need this, delete it.

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        this.loaded = true;
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: What is loaded? Why is this set here?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        console.log(actionType);
+        let date = this.cancelledDate === undefined ? new Date() : this.cancelledDate;
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: You are setting a property in order to pass a date into this function? Why not accept it as input?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+        renewCancel({assetList : this.selectedRows, type: actionType, cancelDate: date})
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: It feels weird to pass a cancellation date in to a renewal call. I guess you did this just to consolidate your controller methods?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+        .then((result) => {
+            
+            console.log('result',result);
+            this.loaded = false
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: I really don't understand this loaded flag. Why do you set it to false here?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        })
+        .catch((error) => {
+            this.loaded = false;
+            console.log('result',error);
+        });
+        this.showCancelPopup = false;
```
Comment (eclapham) [inline force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: What does the popup have to do with this function?

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js [hunk 21]
```diff
@@ -0,0 +1,69 @@
+import { LightningElement, api } from 'lwc';
+import { subscribe, unsubscribe, onError}  from 'lightning/empApi';
+import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+
+export default class PearEventManager extends LightningElement {
+
+    subscription = {};
+     @api channel = '/event/AssetCancelInitiatedEvent';
```
Comment (eclapham) [inline force-app/main/default/lwc/pearEventManager/pearEventManager.js]: This only works for cancel, and not renew?

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+        // Register error listener
+        this.registerErrorListener();
+        this.handleSubscribe();
+    }
+
+    disconnectedCallback() {
+        this.handleUnsubscribe();
+    }
+
+    // Handles subscribe button click
```
Comment (eclapham) [inline force-app/main/default/lwc/pearEventManager/pearEventManager.js]: What button?

File: scripts/apex/hello.apex [hunk 23]
```diff
@@ -0,0 +1,10 @@
+// Use .apex files to store anonymous Apex.
+// You can execute anonymous Apex in VS Code by selecting the
+//     apex text and running the command:
+//     SFDX: Execute Anonymous Apex with Currently Selected Text
+// You can also execute the entire file by running the command:
+//     SFDX: Execute Anonymous Apex with Editor Contents
+
+string tempvar = 'Enter_your_name_here';
```
Comment (eclapham) [inline scripts/apex/hello.apex]: What's the point of this class?

File: scripts/soql/account.soql [hunk 24]
```diff
@@ -0,0 +1,6 @@
+// Use .soql files to store SOQL queries.
+// You can execute queries in VS Code by selecting the
+//     query text and running the command:
+//     SFDX: Execute SOQL Query with Currently Selected Text
+
+SELECT Id, Name FROM Account
```
Comment (eclapham) [inline scripts/soql/account.soql]: Do you need this?

**Review by eclapham — CHANGES_REQUESTED** [review]

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js (hunk 20 above)
Comment (pramodht) [reply force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: This popup used for selecting data for the cancellation.

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js (hunk 21 above)
Comment (pramodht) [reply force-app/main/default/lwc/pearEventManager/pearEventManager.js]: Yes, This is just for the Cancel.

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js (hunk 22 above)
Comment (pramodht) [reply force-app/main/default/lwc/pearEventManager/pearEventManager.js]: I think we can remove the unsubscribe function

File: scripts/apex/hello.apex (hunk 23 above)
Comment (pramodht) [reply scripts/apex/hello.apex]: script folder generated as part of sfdx.. I am deleting it

File: force-app/main/default/classes/AssetManagementControllerTest.cls (hunk 6 above)
Comment (pramodht) [reply force-app/main/default/classes/AssetManagementControllerTest.cls:41]: I will add assert for the inserted data

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js (hunk 20 above)
Comment (eclapham) [reply force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: Yes, but this function itself doesn't seem to interact with the popup, or be tied directly to it? It's just not clear why this belongs here. Maybe I'm just having a hard time following the execution flow of your code, but it feels like this function is used for both renew and cancel, but this popup is only used for cancel, so you're mixing responsibilities. Perhaps this function is trying to do too much, and needs to be broken down?

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js (hunk 21 above)
Comment (eclapham) [reply force-app/main/default/lwc/pearEventManager/pearEventManager.js]: Why don't you want to subscribe to renewal events?

File: force-app/main/default/lwc/pearEventManager/pearEventManager.js (hunk 22 above)
Comment (eclapham) [reply force-app/main/default/lwc/pearEventManager/pearEventManager.js]: You need to clean up any event listeners and subscribers you create, but the comment refers to a button, and I don't think this component has a button? Or am I missing something?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js (hunk 12 above)
Comment (pramodht) [reply force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js:8]: Columns will be visible and the data will not show up for that particular column

File: README.md (hunk 2 above)
Comment (pramodht) [reply README.md:26]: Yes.. this is mandatory, it will fail to have the renewal fields and some fields like hasLifecycle related fields

Comment (eclapham) [reply README.md:26]: You don't have to assign permsets just to deploy code, do you? I wouldn't think so. I'm just saying be very specific with your language here, otherwise you'll just confuse users.

File: force-app/main/default/classes/AssetManagementControllerTest.cls (hunk 6 above)
Comment (eclapham) [reply force-app/main/default/classes/AssetManagementControllerTest.cls:41]: Now with your assertion, this `if` will always pass, so why have it?

File: force-app/main/default/lwc/eventManager/eventManager.js [hunk 25]
```diff
@@ -0,0 +1,77 @@
+import { LightningElement, api } from 'lwc';
+import { subscribe}  from 'lightning/empApi';
+import { ShowToastEvent } from 'lightning/platformShowToastEvent';
+
+export default class EventManager extends LightningElement {
+
+    cancelSubscription = {};
+    renewSubscription = {};
```
Comment (eclapham) [inline force-app/main/default/lwc/eventManager/eventManager.js:10]: Where do you clean these subscribers up?

File: config/project-scratch-def.json [hunk 26]
```diff
@@ -0,0 +1,16 @@
+{
+  "orgName": "Subscription Management",
+  "edition": "Developer",
+  "features": ["EnableSetPasswordInApi", "RevSubscriptionManagement", "OrderSaveLogicEnabled"],
```
Comment (kzheng-sfdc) [inline config/project-scratch-def.json]: What does EnableSetPasswordInApi do? Do you need it?

File: force-app/main/default/classes/AssetManagementController.cls (hunk 5 above)
Comment (kzheng-sfdc) [reply force-app/main/default/classes/AssetManagementController.cls:4]: Not that I know of. Are you referring to the Salesforce copyright header, that kind of stuff?

File: force-app/main/default/classes/AssetManagementController.cls [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+     * Used to call the SM Renew and Cancel service through the Apex via Connect API
+     * @param assetList list of assets to be renewed or cancelled
+     * @param cancelDate Date for the Cancel Service,
+     * @return list of responses from the connectAPI output
+     */
+    
+    @AuraEnabled
+    public static List<String> renewCancelAsset(List<Asset> assetList, String type, Datetime cancelDate){
+        List<String> responseList = new List<String>();
+        ConnectApi.RevenueAsyncRepresentation output;
```
Comment (kzheng-sfdc) [inline force-app/main/default/classes/AssetManagementController.cls:24]: Interesting, I didn't know these are exposed.

File: force-app/main/default/classes/AssetManagementController.cls [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Used to call the SM Renew and Cancel service through the Apex via Connect API
+     * @param assetList list of assets to be renewed or cancelled
+     * @param cancelDate Date for the Cancel Service,
+     * @return list of responses from the connectAPI output
+     */
+    
+    @AuraEnabled
+    public static List<String> renewCancelAsset(List<Asset> assetList, String type, Datetime cancelDate){
```
Comment (kzheng-sfdc) [inline force-app/main/default/classes/AssetManagementController.cls]: Consider breaking this up into two methods?

File: force-app/main/default/classes/AssetManagementController.cls [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+                output = ConnectApi.AssetMgmt.initiateRenewal(input);
+            }
+            else if(type == 'Cancel'){
+                ConnectApi.AssetMgmtInitiateCancellationInput input = new ConnectApi.AssetMgmtInitiateCancellationInput();
+                idList.add(asset.Id);
+                input.assetIds = idList;
+                input.cancellationDate = cancelDate;
+                input.cancellationOutputType = ConnectApi.AssetMgmtCancellationType.Order;
+                output = ConnectApi.AssetMgmt.initiateCancellation(input);
+            }
```
Comment (kzheng-sfdc) [inline force-app/main/default/classes/AssetManagementController.cls]: Throw some kind of exception when type is invalid?

File: force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js (hunk 17 above)
Comment (kzheng-sfdc) [reply force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: When would the canceled date be undefined? Is defaulting to today always OK?

Comment (kzheng-sfdc) [reply force-app/main/default/lwc/assetManagementRelated/assetManagementRelated.js]: Also, I assume the canceled date will be on current user's locale. Do you need to convert it to UTC?

File: force-app/main/default/lwc/eventManager/eventManager.js [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+            // Response contains the subscription information on cancel call
+            this.cancelSubscription = response;
+        });
+
+    }
+
+    
+    handleRenewSubscribe() {
+        // Callback invoked whenever a new event message is received
+        const thisReference = this;
```
Comment (kzheng-sfdc) [inline force-app/main/default/lwc/eventManager/eventManager.js:60]: Use rocket notation (=>) so that you don't need to do thisReference.

File: force-app/main/default/lwc/eventManager/eventManager.js [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+        this.handleRenewSubscribe();
+        this.handleCancelSubscribe();
+    }
+
+    handleCancelSubscribe() {
+        // Callback invoked whenever a new event message is received
+        const thisReference = this;
+        const messageCallback = function(response) {
+            let obj = JSON.parse(JSON.stringify(response));
+            let message = 'Order created: ' + obj.data.payload.CancellationRecordId;
```
Comment (kzheng-sfdc) [inline force-app/main/default/lwc/eventManager/eventManager.js]: You need to match the requestIdentifier before proceeding - otherwise you might be processing events that are not related to your UI.

File: force-app/main/default/lwc/eventManager/eventManager.js [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+
+    }
+
+    
+    handleRenewSubscribe() {
+        // Callback invoked whenever a new event message is received
+        const thisReference = this;
+        const messageCallback = function(response) {
+            let obj = JSON.parse(JSON.stringify(response));
+            let message = 'Order created: ' + obj.data.payload.RenewalRecordId;
```
Comment (kzheng-sfdc) [inline force-app/main/default/lwc/eventManager/eventManager.js]: Same about requestIdentifer here

**Review by kzheng-sfdc — APPROVED** [review]

**Review by kzheng-sfdc — APPROVED** [review]
